// mirador-image-tools/lib/plugins/ImageTool.js
'use strict';

exports.__esModule = true;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _flowRight = require('lodash/flowRight');

var _flowRight2 = _interopRequireDefault(_flowRight);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _MiradorMenuButton = require('mirador/dist/es/src/components/MiradorMenuButton');

var _Slider = require('@material-ui/core/Slider');

var _Slider2 = _interopRequireDefault(_Slider);

var _withStyles = require('@material-ui/core/styles/withStyles');

var _withStyles2 = _interopRequireDefault(_withStyles);

var _withWidth = require('@material-ui/core/withWidth');

var _withWidth2 = _interopRequireDefault(_withWidth);

var _colorManipulator = require('@material-ui/core/styles/colorManipulator');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/** Styles for withStyles HOC */
var styles = function styles(_ref) {
  var _slider;

  var palette = _ref.palette,
      breakpoints = _ref.breakpoints;
  return {
    slider: (_slider = {
      backgroundColor: (0, _colorManipulator.fade)(palette.shades.main, 0.8),
      borderRadius: 25,
      top: 48,
      marginTop: 2,
      position: 'absolute',
      height: 150,
      zIndex: 100,
      marginLeft: 2,
      padding: [[2, 7, 2, 7]]
    }, _slider[breakpoints.down('sm')] = {
      top: 'auto',
      right: 48,
      width: 150,
      height: 'auto',
      marginTop: -46,
      marginBottom: 2,
      padding: [[4, 2, 4, 2]]
    }, _slider)
  };
};

var ImageTool = function (_Component) {
  _inherits(ImageTool, _Component);

  function ImageTool(props) {
    _classCallCheck(this, ImageTool);

    var _this = _possibleConstructorReturn(this, _Component.call(this, props));

    _this.state = {
      open: props.open
    };
    _this.handleChange = _this.handleChange.bind(_this);
    _this.handleClick = _this.handleClick.bind(_this);
    return _this;
  }

  ImageTool.prototype.handleClick = function handleClick() {
    var _props = this.props,
        value = _props.value,
        variant = _props.variant;

    switch (variant) {
      case 'toggle':
        this.handleChange({}, value === 0 ? 100 : 0);
        break;
      default:
        this.setState(function (state) {
          return {
            open: !state.open
          };
        });
    }
  };

  ImageTool.prototype.handleChange = function handleChange(e, val) {
    var onChange = this.props.onChange;

    onChange(val);
  };

  ImageTool.prototype.render = function render() {
    var _props2 = this.props,
        children = _props2.children,
        containerId = _props2.containerId,
        label = _props2.label,
        max = _props2.max,
        min = _props2.min,
        value = _props2.value,
        type = _props2.type,
        variant = _props2.variant,
        windowId = _props2.windowId,
        foregroundColor = _props2.foregroundColor,
        classes = _props2.classes,
        width = _props2.width;
    var open = this.state.open;


    var toggled = variant === 'toggle' && value > 0;

    var id = windowId + '-' + type;

    var bubbleBg = void 0;
    if (open || toggled) {
      bubbleBg = (0, _colorManipulator.fade)(foregroundColor, open ? 0.1 : 0.25);
    }

    return _react2.default.createElement(
      'div',
      { style: { display: 'inline-block' } },
      _react2.default.createElement(
        _MiradorMenuButton.MiradorMenuButton,
        {
          id: id + '-label',
          'aria-label': label,
          containerId: containerId,
          onClick: this.handleClick,
          'aria-expanded': open,
          'aria-controls': id,
          style: { backgroundColor: bubbleBg }
        },
        children
      ),
      open && _react2.default.createElement(
        'div',
        {
          id: id,
          'aria-labelledby': id + '-label',
          className: 'MuiPaper-elevation4 ' + classes.slider
        },
        _react2.default.createElement(_Slider2.default, {
          orientation: ['xs', 'sm'].indexOf(width) >= 0 ? 'horizontal' : 'vertical',
          min: min,
          max: max,
          value: value,
          onChange: this.handleChange
        })
      )
    );
  };

  return ImageTool;
}(_react.Component);

ImageTool.propTypes = process.env.NODE_ENV !== "production" ? {
  children: _propTypes2.default.node.isRequired,
  // eslint-disable-next-line react/forbid-prop-types
  classes: _propTypes2.default.object.isRequired,
  containerId: _propTypes2.default.string.isRequired,
  foregroundColor: _propTypes2.default.string,
  label: _propTypes2.default.string.isRequired,
  min: _propTypes2.default.number,
  max: _propTypes2.default.number,
  onChange: _propTypes2.default.func.isRequired,
  open: _propTypes2.default.bool,
  type: _propTypes2.default.string.isRequired,
  value: _propTypes2.default.number.isRequired,
  variant: _propTypes2.default.string,
  windowId: _propTypes2.default.string.isRequired,
  width: _propTypes2.default.oneOf(['lg', 'md', 'sm', 'xl', 'xs']).isRequired
} : {};

ImageTool.defaultProps = {
  foregroundColor: 'rgb(0, 0, 0)',
  min: 0,
  max: 100,
  open: false,
  variant: 'slider'
};

exports.default = (0, _flowRight2.default)((0, _withStyles2.default)(styles), (0, _withWidth2.default)())(ImageTool);
module.exports = exports['default'];
