// mirador-image-tools/lib/plugins/ImageFlip.js
'use strict';

exports.__esModule = true;
exports.default = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _MiradorMenuButton = require('mirador/dist/es/src/components/MiradorMenuButton');

var _SwapHoriz = require('@material-ui/icons/SwapHoriz');

var _SwapHoriz2 = _interopRequireDefault(_SwapHoriz);

var _colorManipulator = require('@material-ui/core/styles/colorManipulator');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ImageFlip = function (_Component) {
  _inherits(ImageFlip, _Component);

  function ImageFlip() {
    _classCallCheck(this, ImageFlip);

    return _possibleConstructorReturn(this, _Component.apply(this, arguments));
  }

  ImageFlip.prototype.render = function render() {
    var _props = this.props,
        flipped = _props.flipped,
        label = _props.label,
        backgroundColor = _props.backgroundColor,
        foregroundColor = _props.foregroundColor,
        otherProps = _objectWithoutProperties(_props, ['flipped', 'label', 'backgroundColor', 'foregroundColor']);

    return _react2.default.createElement(
      _MiradorMenuButton.MiradorMenuButton,
      _extends({
        'aria-label': label,
        style: { backgroundColor: flipped && (0, _colorManipulator.fade)(foregroundColor, 0.25) }
      }, otherProps),
      _react2.default.createElement(_SwapHoriz2.default, { style: { color: flipped && backgroundColor } })
    );
  };

  return ImageFlip;
}(_react.Component);

exports.default = ImageFlip;


ImageFlip.propTypes = process.env.NODE_ENV !== "production" ? {
  backgroundColor: _propTypes2.default.string,
  containerId: _propTypes2.default.string.isRequired,
  foregroundColor: _propTypes2.default.string,
  flipped: _propTypes2.default.bool.isRequired,
  label: _propTypes2.default.string.isRequired
} : {};

ImageFlip.defaultProps = {
  backgroundColor: 'rgb(255, 255, 255)',
  foregroundColor: 'rgb(0, 0, 0)'
};
module.exports = exports['default'];
