// mirador-image-tools/lib/plugins/MiradorImageToolsMenuItem.js
'use strict';

exports.__esModule = true;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _MenuItem = require('@material-ui/core/MenuItem');

var _MenuItem2 = _interopRequireDefault(_MenuItem);

var _ListItemIcon = require('@material-ui/core/ListItemIcon');

var _ListItemIcon2 = _interopRequireDefault(_ListItemIcon);

var _ListItemText = require('@material-ui/core/ListItemText');

var _ListItemText2 = _interopRequireDefault(_ListItemText);

var _TuneSharp = require('@material-ui/icons/TuneSharp');

var _TuneSharp2 = _interopRequireDefault(_TuneSharp);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var MiradorImageToolsMenuItem = function MiradorImageToolsMenuItem(_ref) {
  var enabled = _ref.enabled,
      handleClose = _ref.handleClose,
      updateWindow = _ref.updateWindow,
      windowId = _ref.windowId;

  var handleClickOpen = function handleClickOpen() {
    handleClose();
    updateWindow(windowId, { imageToolsEnabled: !enabled });
  };

  return _react2.default.createElement(
    _MenuItem2.default,
    { onClick: handleClickOpen },
    _react2.default.createElement(
      _ListItemIcon2.default,
      null,
      _react2.default.createElement(_TuneSharp2.default, null)
    ),
    _react2.default.createElement(
      _ListItemText2.default,
      { primaryTypographyProps: { variant: 'body1' } },
      enabled ? 'Hide image tools' : 'Show image tools'
    )
  );
};

MiradorImageToolsMenuItem.propTypes = process.env.NODE_ENV !== "production" ? {
  enabled: _propTypes2.default.bool,
  handleClose: _propTypes2.default.func.isRequired,
  updateWindow: _propTypes2.default.func.isRequired,
  windowId: _propTypes2.default.string.isRequired
} : {};

MiradorImageToolsMenuItem.defaultProps = {
  enabled: true
};

exports.default = MiradorImageToolsMenuItem;
module.exports = exports['default'];
