// mirador-image-tools/lib/plugins/miradorImageToolsPlugin.js
'use strict';

exports.__esModule = true;

var _withTheme = require('@material-ui/core/styles/withTheme');

var _withTheme2 = _interopRequireDefault(_withTheme);

var _actions = require('mirador/dist/es/src/state/actions');

var actions = _interopRequireWildcard(_actions);

var _selectors = require('mirador/dist/es/src/state/selectors');

var _MiradorImageTools = require('./MiradorImageTools');

var _MiradorImageTools2 = _interopRequireDefault(_MiradorImageTools);

var _MiradorImageToolsMenuItem = require('./MiradorImageToolsMenuItem');

var _MiradorImageToolsMenuItem2 = _interopRequireDefault(_MiradorImageToolsMenuItem);

var _translations = require('../translations');

var _translations2 = _interopRequireDefault(_translations);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = [{
  target: 'OpenSeadragonViewer',
  mapDispatchToProps: {
    updateWindow: actions.updateWindow,
    updateViewport: actions.updateViewport
  },
  mapStateToProps: function mapStateToProps(state, _ref) {
    var windowId = _ref.windowId;
    return {
      containerId: (0, _selectors.getContainerId)(state),
      enabled: (0, _selectors.getWindowConfig)(state, { windowId: windowId }).imageToolsEnabled || false,
      open: (0, _selectors.getWindowConfig)(state, { windowId: windowId }).imageToolsOpen || false,
      viewConfig: (0, _selectors.getViewer)(state, { windowId: windowId }) || {}
    };
  },
  mode: 'add',
  component: (0, _withTheme2.default)(_MiradorImageTools2.default),
  config: {
    translations: _translations2.default
  }
}, {
  target: 'WindowTopBarPluginMenu',
  component: _MiradorImageToolsMenuItem2.default,
  mode: 'add',
  mapDispatchToProps: {
    updateWindow: actions.updateWindow
  },
  mapStateToProps: function mapStateToProps(state, _ref2) {
    var windowId = _ref2.windowId;
    return {
      enabled: (0, _selectors.getWindowConfig)(state, { windowId: windowId }).imageToolsEnabled || false
    };
  }
}];
module.exports = exports['default'];
